README.md
# goodreads_analytics
 


conda-meta/history
==> 2022-05-14 17:01:58 <==
# cmd: /anaconda3/bin/conda create --prefix /Users/hayyu.hanifah/Documents/GitHub/goodreads_analytics python=3.8
# conda version: 4.8.3
+defaults/noarch::wheel-0.37.1-pyhd3eb1b0_0
+defaults/osx-64::ca-certificates-2022.4.26-hecd8cb5_0
+defaults/osx-64::certifi-2021.10.8-py38hecd8cb5_2
+defaults/osx-64::libcxx-12.0.0-h2f01273_0
+defaults/osx-64::libffi-3.3-hb1e8313_2
+defaults/osx-64::ncurses-6.3-hca72f7f_2
+defaults/osx-64::openssl-1.1.1o-hca72f7f_0
+defaults/osx-64::pip-21.2.4-py38hecd8cb5_0
+defaults/osx-64::python-3.8.13-hdfd78df_0
+defaults/osx-64::readline-8.1.2-hca72f7f_1
+defaults/osx-64::setuptools-61.2.0-py38hecd8cb5_0
+defaults/osx-64::sqlite-3.38.3-h707629a_0
+defaults/osx-64::tk-8.6.11-h3fd3227_1
+defaults/osx-64::xz-5.2.5-hca72f7f_1
+defaults/osx-64::zlib-1.2.12-h4dc903c_2
# update specs: ['python=3.8']


goodreads_spider.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun May  1 22:25:45 2022

@author: hayyu.hanifah
"""

# Import scrapy
import scrapy

# Import the CrawlerProcess
from scrapy.crawler import CrawlerProcess

# Create the Spider class
class YourSpider(scrapy.Spider):
  name = 'yourspider'
  # start_requests method
  def start_requests( self ):
    url_short = "https://www.goodreads.com/notes/38648728-hayyu-hanifah"
    yield scrapy.Request(url = url_short, callback=self.parse)
      
  def parse(self, response):
    # Get the link to the highlight items of each book
    book_link_list = response.xpath('//a[@class="annotatedBookItem__knhLink")]/@href/text()').extract()
    print(len(book_link_list))
    for bl in book_link_list:
      yield response.follow(url = bl, callback = self.parse_highlight)
  
  def parse_highlight(self, response):
    highlight_url = response.url
    print(highlight_url)
    highlight_list = response.xpath('//div[@class="noteHighlightTextContainer__highlightText"]/span/text()').extract()
    print(len(highlight_list))
    gr_dict[highlight_url] = highlight_list
    
# Initialize the dictionary **outside** of the Spider class
gr_dict = dict()

# Run the Spider
process = CrawlerProcess()
process.crawl(YourSpider)
process.start()

# Print a preview of courses
print(gr_dict)

test_response = scrapy.Request(url = "https://www.goodreads.com/notes/38648728-hayyu-hanifah")._get_body()
print(test_response.url)
book_link_list = test_response.xpath('//a[contains(@class,"annotatedBookItem__knhLink")]/@href/text()').extract()
print(len(book_link_list))
for bl in book_link_list:
  highlight_response = test_response.follow(url = bl)
  highlight_url = highlight_response.url
  print(highlight_url)
  highlight_list = highlight_response.xpath('//div[@class="noteHighlightTextContainer__highlightText"]/span/text()').extract()
  print(len(highlight_list))
  

goodreads_analytics.py
# -*- coding: utf-8 -*-
"""
Goodreads Analytics

AUTHOR: Hayyu Hanifah

An interactive dashboard to display Goodreads data. Built using Streamlit.

Metrics:
    1. Number of books read (overall, breakdown by fiction/non-fiction, by genre)
    2. Number of page read (overall, breakdown by fiction/non-fiction, by genre)
    3. Number of books added to TBR (overall, breakdown by fiction/non-fiction, by genre)
    4. Reading velocity over the year (what month I read the most, what genre)
    5. Highlights -> cloud of words (to get the vibes of what interest me the most)
    6. My ratings (overall and per month); my rating vs overall rating of the book (is it usually similar, higher, or lower)
    
Features:
    1. Data visualization using plotly (+ filtering)
    2. Upload data (to enable others if they want to visualize their Goodreads data)
    3. Crawling Goodreads to get the book info (genre, highlights)
    4. Recommend books to read
    
TODO:
    1. Data transformation (make sure date type of date read, date added; numeric and string types)
    2. Get the highlights of read books (from kindle highlight, need to learn about selenium) -> should be separated from the main streamlit app, since it needs credentials 
    3. Visualisation (create scorecards, bar charts, line charts, pie charts, word cloud)
    
"""

import pandas as pd 
import numpy as np 
import plotly.graph_objects as go
import plotly.express as px
from plotly.subplots import make_subplots
import streamlit as st
from datetime import datetime
import json
import os
import warnings
warnings.filterwarnings('ignore') ## to avoid error with keyerror 'warnings'

### FUNCTIONS
def get_dict_from_file(filepath):
    with open(filepath, 'r') as file:
        json_str = file.read()

    json_dict = json.loads(json_str)
    return json_dict


### STREAMLIT
working_dir_path = os.environ['CONDA_PREFIX']
param = get_dict_from_file(os.path.join(working_dir_path,'parameters.json'))

@st.cache
def load_data(param):
    """ Loads in 4 dataframes and does light feature engineering"""
    df = pd.read_csv(param['consolidated_data_filepath'])
    
    # Data Preparation
    date_cols = ['date_read','date_added','add_to_tbr_dt','start_reading_dt','finish_reading_dt']
    for c in date_cols:
        df[c] = pd.to_datetime(df[c])
    
    df['add_to_tbr_dt'] = df.apply(lambda x: x['date_added'] if pd.isna(x['add_to_tbr_dt']) else x['add_to_tbr_dt'], axis=1)
    df['wtr_to_start_read'] = (df['start_reading_dt'] - df['add_to_tbr_dt']).dt.days
    df['start_to_finish_read'] = (df['date_read'] - df['start_reading_dt']).dt.days
    return df

# Data Aggregation
main_df = load_data(param)

## Conditions
null_date_read = main_df['date_read'].isna()
null_date_start = main_df['start_reading_dt'].isna()
fiction = (main_df['is_fiction'] == 1)
nonfiction = (main_df['is_nonfiction'] == 1)
unknown_category = (main_df['is_fiction'] == 0) & (main_df['is_nonfiction'] == 0)
# read = (main_df['Exclusive Shelf'] == 'read') | (main_df['Read Count'] > 0)
read = (main_df['Exclusive Shelf'] == 'read')

## Overall
total_books = main_df['Book Id'].nunique()
total_books_read = main_df[~null_date_read | read]['Book Id'].nunique()
total_books_inprogress = main_df[(~null_date_start) & (null_date_read)]['Book Id'].nunique()

### Fiction
total_books_fic = main_df[fiction]['Book Id'].nunique()
total_books_read_fic = main_df[(~null_date_read | read) & fiction]['Book Id'].nunique()
total_books_inprogress_fic = main_df[(~null_date_start) & (null_date_read) & fiction]['Book Id'].nunique()

### Non Fiction
total_books_nonfic = main_df[nonfiction]['Book Id'].nunique()
total_books_read_nonfic = main_df[(~null_date_read) & nonfiction]['Book Id'].nunique()
total_books_inprogress_nonfic = main_df[(~null_date_start) & (null_date_read) & nonfiction]['Book Id'].nunique()

### Unknown Category
total_books_unknown = main_df[unknown_category]['Book Id'].nunique()
total_books_read_unknown = main_df[(~null_date_read) & unknown_category]['Book Id'].nunique()
total_books_inprogress_unknown = main_df[(~null_date_start) & (null_date_read) & unknown_category]['Book Id'].nunique()

overall_by_category = pd.DataFrame(
                        data={
                                "category": ['fiction','non-fiction','unknown'],
                                "total_books": [total_books_fic, total_books_nonfic, total_books_unknown],
                                "total_books_read": [total_books_read_fic, total_books_read_nonfic, total_books_read_unknown],
                                "total_books_inprogress": [total_books_inprogress_fic, total_books_inprogress_nonfic, total_books_inprogress_unknown]
                            }
    )


## Yearly


## Monthly


# Visualization
st.title("Goodreads Analytics")
st.markdown("""
            ## Overview
            Total number of books on my shelves as per 1 May 2022, how many have been read,
            and how many still not finish reading.
            """)
labels = overall_by_category['category']
colors = px.colors.qualitative.Pastel
fig_total_books = make_subplots(rows=1, cols=3, 
                                subplot_titles=("Total Books","Read","In Progress"),
                                specs=[[{'type':'domain'}, {'type':'domain'}, {'type':'domain'}]],
                                
                                )
fig_total_books.add_trace(go.Pie(labels=labels, values=overall_by_category['total_books'], 
                                 name="Total Books",
                                 insidetextorientation='horizontal'
                                 ), 
                          1, 1)
fig_total_books.add_trace(go.Pie(labels=labels, values=overall_by_category['total_books_read'], 
                                 name="Books Read",
                                 insidetextorientation='horizontal'
                                 ), 
                          1, 2)
fig_total_books.add_trace(go.Pie(labels=labels, values=overall_by_category['total_books_inprogress'], 
                                 name="Books In Progress",
                                 insidetextorientation='horizontal'
                                 ), 
                          1, 3)
fig_total_books.update_traces(hole=.52, 
                              hoverinfo="label+percent+value",
                              marker=dict(colors=colors[2:5])
                              )

## Tricks to find the coordinate for center annotations
# coord_annot = []
# for p in list(np.arange(start=0, stop=1.1, step=0.1)):
#     current_step = "{:.1f}".format(p)
#     coord_annot.append(dict(text=current_step, x=p, y=0, font_size=15, showarrow=False))
#     coord_annot.append(dict(text=current_step, x=0, y=p, font_size=15, showarrow=False))
    
# donut_existing_annot += coord_annot

donut_center_annot = [
        dict(text=str(total_books), x=0.103, y=0.53, font_size=27, showarrow=False),
        dict(text="books", x=0.103, y=0.43, font_size=15, showarrow=False),
        dict(text=str(total_books_read), x=0.5, y=0.53, font_size=27, showarrow=False),
        dict(text="books", x=0.5, y=0.43, font_size=15, showarrow=False),
        dict(text=str(total_books_inprogress), x=0.885, y=0.53, font_size=27, showarrow=False),
        dict(text="books", x=0.895, y=0.43, font_size=15, showarrow=False)
    ]

donut_existing_annot = [a.to_plotly_json() for a in fig_total_books["layout"]["annotations"]]
donut_existing_annot += donut_center_annot # so the center annot and subplot title not override each other

fig_total_books.update_layout(
        annotations=donut_existing_annot,
        legend=dict(yanchor="middle",y=0.5),
        margin=dict(l=3, r=3, t=0, b=3),
        autosize=False,
        height=250
    )
st.plotly_chart(fig_total_books)



goodreads_beautifulsoup.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon May  2 10:04:43 2022

@author: hayyu.hanifah
"""

from bs4 import BeautifulSoup
import requests
from lxml import etree


url = "https://www.goodreads.com/notes/38648728-hayyu-hanifah"
soup = BeautifulSoup(requests.get(url).content, "lxml")
dom = etree.HTML(str(soup))
# print(dom)
cek = dom.xpath('/html/body/div[1]/div[3]/div[1]/div[1]/div[2]/div/div/div/div[2]/div/div[2]')
print(cek)
# book_page_html = soup.find_all('a', class_="annotatedBookItem__knhLink") #class_="annotatedBookItem__knhLink"

# book_page_html = soup.select("a.annotatedBookItem__knhLink")

# annot = book_page_html.find('a', class_=True, recursive=True)
# annot_list = [x for x in annot if x.attrs['class'] == "annotatedBookItem__knhLink"]
# urls = [x['href'] for x in links]
# print(urls)
# print(links)

# classes = [value 
#             for element in annot 
#             for value in element["class"]]
# print(classes)

# print(book_page_html)
# annot[10].attrs['class']

# print(annot)

goodreads_scraper.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun May  1 19:50:17 2022

@author: hayyu.hanifah
"""

from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time

from bs4 import BeautifulSoup
import requests

import pandas as pd
import json
import os

def selenium_find_elements(driver, url, by_id, by_value):
    driver.get(url)
    time.sleep(10)
    if by_id == 'class':
        e_list = driver.find_elements(by=By.CLASS_NAME, value=by_value)
    # TODO: implement find elements by other identifiers
    return e_list

def bs_find_all(url, tag_string, attrs, headers=''):
    if headers == '':
        page_source = requests.get(url).content
    else:
        page_source = requests.get(url, headers=headers).content
    page_soup = BeautifulSoup(page_source, "lxml")
    elements = page_soup.find_all(tag_string, attrs=attrs)
    return elements

def get_dict_from_file(filepath):
    with open(filepath, 'r') as file:
        json_str = file.read()

    json_dict = json.loads(json_str)
    return json_dict

def convert_str_list(row_df,str_col):
    clean_str = row_df[str_col].replace('[','').replace(']','').replace("'","")
    str_list = clean_str.split(", ")
    return str_list

def get_timeline_date(row_df,str_col,substr_timeline):
    current_timelines = row_df[str_col]
    dt_list = []
    for t in current_timelines:
        if substr_timeline in t:
            dt = t.replace(substr_timeline,'').strip()
            dt_list.append(dt)
    return dt_list

def check_is_genre(row_df,str_col,genre_str):
    current_genre_list = row_df[str_col]
    if genre_str in current_genre_list:
        return 1
    else:
        return 0
    
def exclude_genre(row_df,str_col,genre_str):
    current_genre_list = row_df[str_col]
    genre = [x for x in current_genre_list if x != genre_str]
    return genre

# LOAD PARAMETERS FILE
working_dir_path = os.environ['CONDA_PREFIX']
param = get_dict_from_file(os.path.join(working_dir_path,'parameters.json'))

# Setup Selenium
s = Service(param['chromedriver_path'])
driver = webdriver.Chrome(service=s)

# Get the list of books with public highlights
highlight_lp_url = param['highlight_lp_url']
book_list = selenium_find_elements(driver, highlight_lp_url, 'class', 'annotatedBookItem__knhLink')
print(len(book_list))

highlight_dict = dict()

for l in book_list:
    # For each book, get the list of hightlighted sentences
    current_h_link = l.get_attribute("href")
    print(current_h_link)
    highlight_list = bs_find_all(current_h_link, 'div', attrs={"class":"noteHighlightTextContainer__highlightText"})
    hl_text_list = []
    
    for hl in highlight_list:
        print(hl.find('span').text)
        hl_text_list.append(hl.find('span').text)
    
    # highlight_dict[current_h_link] = hl_text_list
    highlight_dict[l] = hl_text_list
    
print(highlight_dict)
with open(param['highlight_filepath'], 'w') as convert_file:
      convert_file.write(json.dumps(highlight_dict))

# Get the top shelves of each book from Genre section
options = Options()
options.headless = True
book_driver = webdriver.Chrome(service=s,options=options)
executor_url = book_driver.command_executor._url
session_id = book_driver.session_id

print(session_id)
print(executor_url)

book_df = pd.read_csv(param['gr_library_export_filepath'])
book_id_list = book_df['Book Id'].tolist()
url_prefix = 'https://www.goodreads.com/book/show/'

books_dict = dict()
for bid in book_id_list[100:312]:  ## crawl the genre in batches
    print(bid)
    current_href = url_prefix + str(bid)
    book_driver.get(current_href)
    time.sleep(7)
    parent = book_driver.current_window_handle
    uselessWindows = book_driver.window_handles
    for winId in uselessWindows:
        if winId != parent: 
            book_driver.switch_to.window(winId)
            book_driver.close()
            
    book_soup = BeautifulSoup(book_driver.page_source, "html.parser")
    genre_list = book_soup.find_all('div',attrs={'class':'elementList'})
    print([x.find('a').text for x in genre_list])
    books_dict[str(bid)] = [x.find('a').text for x in genre_list]

with open(param['genre_filepath_3'], 'w') as convert_file:
      convert_file.write(json.dumps(books_dict))

# Concat book genre files
genre_1 = get_dict_from_file(param['genre_filepath_1'])
genre_2 = get_dict_from_file(param['genre_filepath_2'])
genre_3 = get_dict_from_file(param['genre_filepath_3'])

genre_1.update(genre_2) # it is safe since there are no overlap keys between 2 dicts
genre_1.update(genre_3)

print(len(genre_1))
print(genre_1)

book_id_list = [k for k,v in genre_1.items()]
book_genre_list = [v for k,v in genre_1.items()]

genre_df = pd.DataFrame.from_dict({'book_id': book_id_list, 'book_genre': book_genre_list})

def clean_genre_list(row_df, genre_col):
    raw_list = [x for x in row_df[genre_col] if x != '']
    return sorted(set(raw_list), key=raw_list.index)

genre_df['clean_genre_list'] = genre_df.apply(lambda r: clean_genre_list(r, 'book_genre'), axis=1)
    
genre_df.to_csv(param['clean_genre_filepath'],index=False)

# Get book timeline details 
## Get review list, to get the link to review detail pages
review_driver = webdriver.Chrome(service=s)
review_driver.get(param['review_lp_url'])
time.sleep(7)
parent = book_driver.current_window_handle
uselessWindows = book_driver.window_handles
for winId in uselessWindows:
    if winId != parent: 
        book_driver.switch_to.window(winId)
        book_driver.close()
for i in range(15): # simulate scroll, to load all the review items
    review_driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")
    time.sleep(4)

review_soup = BeautifulSoup(review_driver.page_source, "html.parser")
review_list = review_soup.find_all('td',attrs={'class':'field actions'})
review_link_list = [x.find('a').attrs['href'] for x in review_list]

print(review_link_list)

review_url_prefix = 'https://www.goodreads.com'
timeline_dict = dict()
for r in review_link_list:
    book_driver.get(review_url_prefix+r)
    time.sleep(7)
    parent = book_driver.current_window_handle
    uselessWindows = book_driver.window_handles
    for winId in uselessWindows:
        if winId != parent: 
            book_driver.switch_to.window(winId)
            book_driver.close()
    
    book_soup = BeautifulSoup(book_driver.page_source, "html.parser")
    book_title = book_soup.find('h1').text.replace("Hayyu Hanifah's Reviews > ","")
    book_timeline = book_soup.find_all('div', attrs={'class':'readingTimeline__text'})
    book_timeline_list = [x.text.strip().replace('\n',' ') for x in book_timeline]
    timeline_dict[book_title] = book_timeline_list
    
print(timeline_dict)
with open(param['timeline_filepath'], 'w') as convert_file:
      convert_file.write(json.dumps(timeline_dict))

### CONSOLIDATE ALL DATASET
goodreads_lib_export = pd.read_csv(param['gr_library_export_filepath'])
book_genre_df = pd.read_csv(param['clean_genre_filepath'])
timeline = get_dict_from_file(param['timeline_filepath'])

consolidate_df = goodreads_lib_export.merge(book_genre_df, how='left', left_on='Book Id', right_on='book_id')
consolidate_df.drop(columns=['book_id', 'book_genre'],inplace=True)

consolidate_df['date_read'] = pd.to_datetime(consolidate_df['Date Read'])
consolidate_df['date_added'] = pd.to_datetime(consolidate_df['Date Added'])
consolidate_df['Title_shorten'] = consolidate_df.apply(lambda x: x['Title'].split(" (")[0].strip() if x['Title'].endswith(")") else x['Title'].strip(), axis=1)

book_title = [k for k,v in timeline.items()]
timelines = [v for k,v in timeline.items()]
book_timeline_df = pd.DataFrame(data={'book_title': book_title, 'timelines': timelines})
book_timeline_df['add_to_tbr'] = book_timeline_df.apply(lambda x: get_timeline_date(x,'timelines','– Shelved as: to-read'), axis=1)
book_timeline_df['start_reading'] = book_timeline_df.apply(lambda x: get_timeline_date(x,'timelines','–  Started Reading'), axis=1)
book_timeline_df['finish_reading'] = book_timeline_df.apply(lambda x: get_timeline_date(x,'timelines','–  Finished Reading'), axis=1)

book_timeline_df['add_to_tbr_dt_len'] = book_timeline_df.apply(lambda x: len(x['add_to_tbr']),axis=1)
book_timeline_df['start_reading_dt_len'] = book_timeline_df.apply(lambda x: len(x['start_reading']),axis=1)
book_timeline_df['finish_reading_dt_len'] = book_timeline_df.apply(lambda x: len(x['finish_reading']),axis=1)

book_timeline_df[book_timeline_df['finish_reading_dt_len'] > 1]

# df.A.map(lambda x: x[0])
book_timeline_df['add_to_tbr_dt'] = book_timeline_df.add_to_tbr.map(lambda x: x[0] if len(x) == 1 else '')
book_timeline_df['start_reading_dt'] = book_timeline_df.start_reading.map(lambda x: x[0] if len(x) == 1 else '')
book_timeline_df['finish_reading_dt'] = book_timeline_df.finish_reading.map(lambda x: x[0] if len(x) == 1 else '')

book_timeline_df.drop(columns=['add_to_tbr','start_reading','finish_reading','add_to_tbr_dt_len','start_reading_dt_len','finish_reading_dt_len'],inplace=True)

consolidate_df = consolidate_df.merge(book_timeline_df, how='left', left_on='Title_shorten', right_on='book_title')

consolidate_df['add_to_tbr_dt'] = pd.to_datetime(consolidate_df['add_to_tbr_dt'], infer_datetime_format=True)
consolidate_df['start_reading_dt'] = pd.to_datetime(consolidate_df['start_reading_dt'], infer_datetime_format=True)
consolidate_df['finish_reading_dt'] = pd.to_datetime(consolidate_df['finish_reading_dt'], infer_datetime_format=True)

consolidate_df.dropna(axis=1, how='all', inplace=True) # drop all columns which only consist of NaN value

consolidate_df['genre_list'] = consolidate_df.apply(lambda x: convert_str_list(x,'clean_genre_list'), axis=1)

consolidate_df['is_nonfiction'] = consolidate_df.apply(lambda x: check_is_genre(x,'genre_list','Nonfiction'), axis=1)
consolidate_df['is_fiction'] = consolidate_df.apply(lambda x: check_is_genre(x,'genre_list','Fiction'), axis=1)

consolidate_df['genre_list_trf'] = consolidate_df.apply(lambda x: exclude_genre(x,'genre_list','Nonfiction'), axis=1)
consolidate_df['genre_list_trf'] = consolidate_df.apply(lambda x: exclude_genre(x,'genre_list_trf','Fiction'), axis=1)

consolidate_df['year_read'], consolidate_df['month_read'] = consolidate_df['date_read'].dt.year, consolidate_df['date_read'].dt.month
consolidate_df['year_added'], consolidate_df['month_added'] = consolidate_df['date_added'].dt.year, consolidate_df['date_added'].dt.month

consolidate_df.to_csv(param['consolidated_data_filepath'], index=False)

